internal/handler/favourites.go

```diff
@@ -0,0 +1,296 @@
+package handler
+
+import (
+	"encoding/json"
+	"net/http"
+	"strconv"
+	"sync"
+
+	"svenvermeulen/platform-go-challenge/internal/auth"
+	"svenvermeulen/platform-go-challenge/internal/repository/audience"
+	"svenvermeulen/platform-go-challenge/internal/repository/chart"
+	"svenvermeulen/platform-go-challenge/internal/repository/favourite"
+	"svenvermeulen/platform-go-challenge/internal/repository/insight"
+	"svenvermeulen/platform-go-challenge/pkg/model"
+
+	"github.com/gin-gonic/gin"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+type FavouritesHandler struct {
+	favouriteRepository *favourite.Repository
+	audienceRepository  *audience.Repository
+	chartRepository     *chart.Repository
+	insightRepository   *insight.Repository
+}
+
+func NewFavouritesHandler(favouriteRepository *favourite.Repository,
+	audienceRepository *audience.Repository,
+	chartRepository *chart.Repository,
+	insightRepository *insight.Repository) *FavouritesHandler {
+	return &FavouritesHandler{
+		favouriteRepository: favouriteRepository,
+		audienceRepository:  audienceRepository,
+		chartRepository:     chartRepository,
+		insightRepository:   insightRepository,
+	}
+}
+
+// GetUserFavourites godoc
+//
+// @Summary     Get user favourites
+// @Description gets list of favourites for a specified user
+// @Tags        favourites
+// @Produce     json
+// @Success     200	{object}	[]model.UserFavourite
+// @Failure     400
+// @Failure     401
+// @Failure     404
+// @Failure     500
+// @Router      /favourites/:userid [get]
+func (h *FavouritesHandler) GetFavourites(c *gin.Context) {
+	userId, err := auth.GetUserIDFromToken(c)
+	if err != nil {
+		log.Errorf("Error obtaining userid from jwt token: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	log.Info("Retrieving favourites for userid", userId)
+
+	var offset int
+	var pageSize int
+
+	if offset, err = strconv.Atoi(c.Query("offset")); err != nil {
+		offset = 0
+	}
+	if pageSize, err = strconv.Atoi(c.Query("pagesize")); err != nil {
+		pageSize = 10
+	}
+
+	userFavourites := h.favouriteRepository.GetFavourites(userId, offset, pageSize)
+
+	// Get favourite items for current user and extract these into 3 slices of IDs
+	// IDs are then used to query the various repositories
+	// Then I wait for all the queries to return and stitch the results back together
+	audienceIDs, insightIDs, chartIDs := h.splitUserFavourites(userFavourites)
+
+	var wg sync.WaitGroup
+	wg.Add(3)
+
+	// AUDIENCES
+	var audiences map[uuid.UUID]*model.Audience
+
+	go func() {
```

Comment:
Nice using concurrency and wg !

Reply:
@loukaspe Thanks for reviewing, I added my answers to your comments and questions on the PR. Would be happy to discuss more. Cheers.

Reply:
Thank you for answering these extra questions. Appreciate it !